src/java/org/apache/cassandra/cql3/Vectors.java

```diff
@@ -0,0 +1,176 @@
+    }
+
+    private static ColumnSpecification valueSpecOf(ColumnSpecification column)
+    {
+        return new ColumnSpecification(column.ksName, column.cfName, new ColumnIdentifier("value(" + column.name + ")", true), elementsType(column.type));
+    }
+
+    public static class Literal extends Term.Raw
+    {
+        private final List<Term.Raw> elements;
+
+        public Literal(List<Term.Raw> elements)
+        {
+            this.elements = elements;
+        }
+
+        @Override
+        public TestResult testAssignment(String keyspace, ColumnSpecification receiver)
+        {
+            if (!(receiver.type instanceof VectorType))
+                return AssignmentTestable.TestResult.NOT_ASSIGNABLE;
+            VectorType<?> type = (VectorType<?>) receiver.type;
+            if (elements.size() != type.dimension)
+                return AssignmentTestable.TestResult.NOT_ASSIGNABLE;
+            ColumnSpecification valueSpec = valueSpecOf(receiver);
+            return AssignmentTestable.TestResult.testAll(receiver.ksName, valueSpec, elements);
+        }
+
+        @Override
+        public Term prepare(String keyspace, ColumnSpecification receiver) throws InvalidRequestException
+        {
+            if (!(receiver.type instanceof VectorType))
+                throw new InvalidRequestException(String.format("Invalid vector literal for %s of type %s", receiver.name, receiver.type.asCQL3Type()));
+            VectorType<?> type = (VectorType<?>) receiver.type;
+            if (elements.size() != type.dimension)
+                throw new InvalidRequestException(String.format("Invalid vector literal for %s of type %s; expected %d elements, but given %d", receiver.name, receiver.type.asCQL3Type(), type.dimension, elements.size()));
+
+            ColumnSpecification valueSpec = valueSpecOf(receiver);
+            List<Term> values = new ArrayList<>(elements.size());
+            boolean allTerminal = true;
+            for (Term.Raw rt : elements)
+            {
+                if (!rt.testAssignment(keyspace, valueSpec).isAssignable())
+                    throw new InvalidRequestException(String.format("Invalid vector literal for %s: value %s is not of type %s", receiver.name, rt, valueSpec.type.asCQL3Type()));
+
+                Term t = rt.prepare(keyspace, valueSpec);
+
+                if (t instanceof Term.NonTerminal)
+                    allTerminal = false;
+
+                values.add(t);
+            }
+            DelayedValue<?> value = new DelayedValue<>(type, values);
+            return allTerminal ? value.bind(QueryOptions.DEFAULT) : value;
+        }
+
+        @Override
+        public String getText()
+        {
+            return Lists.listToString(elements, Term.Raw::getText);
+        }
+
+        @Override
+        public AbstractType<?> getExactTypeIfKnown(String keyspace)
+        {
+            // not enough information to know dimension
```

Comment:
I wrote the following to match `ListType` but feel this is not correct

```
@Override
        public AbstractType<?> getExactTypeIfKnown(String keyspace)
        {
            // TODO - this doesn't feel right... if you are dealing with a literal then the value is `null`, so we will ignore
            // if there are multiple times, we randomly select the first?  This logic matches Lists.getExactListTypeIfKnown but feels flawed
            Optional<? extends AbstractType<?>> opt = elements.stream()
                                                              .map(e -> e.getCompatibleTypeIfKnown(keyspace))
                                                              .filter(Objects::nonNull)
                                                              .findFirst();
            return opt.isPresent() ? VectorType.getInstance(opt.get(), elements.size()) : null;
        }

        @Override
        public AbstractType<?> getCompatibleTypeIfKnown(String keyspace)
        {
            // TODO - this doesn't feel right... if you are dealing with a literal then the value is `null`, so we will ignore
            // if there are multiple times, we randomly select the first?  This logic matches Lists.getExactListTypeIfKnown but feels flawed
            Set<AbstractType<?>> types = elements.stream()
                                                           .map(e -> e.getCompatibleTypeIfKnown(keyspace))
                                                           .filter(Objects::nonNull)
                                                           .collect(Collectors.toSet());
            AbstractType<?> type = AssignmentTestable.getCompatibleTypeIfKnown(types);
            return type == null ? null : VectorType.getInstance(type, elements.size());
        }
```

I am not sure how best to test these interfaces to make sure they do the right thing in all context, so hard to say... I really hate that we filter `null`, and really don't like `getExactTypeIfKnown`'s `findFirst` as we could have multiple types!!!

@adelapena know how best to test these impl?

Reply:
your test doesn't test this, as it works with `Selectable`; so never hits this logic (as `Selectable` also only works with `ListType`)

Reply:
One big reason why I don't want to add `Selectable` in this patch is that we can't tell if we want `ListType` or `VectorType` without changing a few methods, and adding support can break existing functions that rely on `ListType`

As far as I can tell, this is only a problem for new functions (which are not in this ticket) that work with literals rather than data from a table... so feel its best to leave to the PR that adds those functions....

Reply:
bit the bullet and implemented... I made sure `token` now works for all types